src/coreclr/System.Private.CoreLib/src/Microsoft/Win32/OAVariantLib.cs

```diff
@@ -32,96 +34,196 @@ internal static unsafe partial class OAVariantLib
+            {
+                if (source.GetType() == typeof(int) || source.GetType() == typeof(uint))
+                {
+                    uint sourceData = source.GetType() == typeof(int) ? (uint)(int)source : (uint)source;
+                    // Int32/UInt32 can be converted to System.Drawing.Color
+                    ConvertOleColorToSystemColor(ObjectHandleOnStack.Create(ref result), sourceData, targetClass.TypeHandle.Value);
+                    Debug.Assert(result != null);
+                    return result;
+                }
+            }
 
-#pragma warning restore CS8500
+            VarEnum vt = GetVTFromClass(targetClass);
 
-        #endregion
+            ComVariant vOp = ToOAVariant(source);
+            ComVariant ret = default;
 
+            int hr = Interop.OleAut32.VariantChangeTypeEx(&ret, &vOp, culture.LCID, options, (ushort)vt);
 
-        #region Private Helpers
+            using (vOp)
+            using (ret)
+            {
+                if (hr < 0)
+                    OAFailed(hr);
 
-        private static int GetCVTypeFromClass(Type ctype)
-        {
-            Debug.Assert(ctype != null);
-            Debug.Assert(ClassTypes[CV_OBJECT] == typeof(object), "OAVariantLib::ClassTypes[CV_OBJECT] == Object.class");
+                result = FromOAVariant(ret);
+                if (targetClass == typeof(char))
+                {
+                    result = (char)(uint)result;
+                }
+            }
 
-            // OleAut Binder works better if unrecognized
-            // types were changed to Object.
-            int cvtype = CV_OBJECT;
+            return result;
+        }
 
-            for (int i = 0; i < ClassTypes.Length; i++)
+        private static void OAFailed(int hr)
+        {
+            switch (hr)
             {
-                if (ctype.Equals(ClassTypes[i]))
-                {
-                    cvtype = i;
-                    break;
-                }
+                case HResults.COR_E_OUTOFMEMORY:
+                    throw new OutOfMemoryException();
+                case unchecked((int)0x80020008): // DISP_E_BADVARTYPE
+                    throw new NotSupportedException(SR.NotSupported_OleAutBadVarType);
+                case HResults.COR_E_DIVIDEBYZERO:
+                    throw new DivideByZeroException();
+                case HResults.COR_E_OVERFLOW:
+                    throw new OverflowException();
+                case HResults.TYPE_E_TYPEMISMATCH:
+                    throw new InvalidCastException(SR.InvalidCast_OATypeMismatch);
+                case HResults.E_INVALIDARG:
+                    throw new ArgumentException();
+                default:
+                    Debug.Fail("Unrecognized HResult - OAVariantLib routine failed in an unexpected way!");
+                    throw Marshal.GetExceptionForHR(hr);
             }
+        }
+
+        private static ComVariant ToOAVariant(object input)
```

Comment:
I need to understand more about `OAVaraintLib` and `OleAutBinder`. This path does not use the same variant marshal methods and does not handle some cases like wrapped VT types. How indeed would this path be invoked?

Reply:
It's called by CLRToCOMLateBoundWorker. Perhaps we should make it's behavior consistent with other COM interop scenario? This PR is trying keep things as-is.

Reply:
This is legacy COM interop support. We are not interested in changing any observable behavior. We have learned hard way multiple times that the applications out there depend on every observable behavior detail of legacy COM interop.

Reply:
Well how about the fix of misplacement in CV_TYPE array? Should such bug-to-bug behavior be preserved? Did it fail anyway in caller?

Reply:
If there are cases that are 100% bugs - like something that would always crash and there is no way that a working code can depend on it, it should be ok to fix those. We should have a test case added for any cases like that.

Reply:
Compatibility table to check:

|Type|CVTypes|OleAutBinder behavior|
|-|-|-|
|Primitive numbers|Corresponding CVType|Corresponding VT|
|Enums|CV_ENUM with enum flags|NotSupported in CVToVT|
|IntPtr/UIntPtr|CV_OBJECT|Boxed as IUnknown|
|string|CV_STR|VT_BSTR|
|Empty/DBNull/Missing|Corresponding CVType|Corresponding VT|
|Array|CV_OBJECT with mask|Boxed as IUnknown|
|Wrappers|CV_OBJECT with VT mask|Boxed as IUnknown|
|System.Drawing.Color|CV_U4|VT_UI4|
|decimal/DateTime|CV_DECIMAL/CV_DATETIME|VT_DECIMAL/VT_DATETIME|
|TimeSpan|CV_TIMESPAN|NotSupported in CVToVT|
|bool|CV_BOOLEAN|VT_BOOLEAN|
|char|CV_CHAR|VT_BSTR|

I'm not sure where's the test that exercises OleAutBinder.

Reply:
Well what about the behavior of new source-generated COM interop? While trying to reuse the conversion logic, I noticed a difference that legacy COM would convert VT_ERROR with DISP_E_PARAMNOTFOUND to Missing instead of int. Should we change the behavior (of new mechanism) to align, or keep the inconsistency forever? /cc @jkoritzinsky

Reply:
> Should we change the behavior (of new mechanism) to align, or keep the inconsistency forever?

Sadly yes. The built-in behavior is "correct" and should remain as it helps porting from .NET Framework to .NET  8+. Once teams are on .NET 8+, the highest priority, they can then port to the source generated solution with the consistent COM behavior. As @jkotas has mentioned the built-in COM behavior should remain static and unchanged whenever possible. Porting large products from .NET Framework are impacted enough by newer .NET semantics as in and forcing them to also take on COM interop semantics changes is a burden we _need_ to avoid - see https://github.com/dotnet/runtime/pull/100377 for an example affordance we are considering.

Reply:
I mean that changing the *other* routines to match built-in behavior. Specifically speaking, `BuiltInInteropVariantExtensions.ToObject` is missing a case of `VT_ERROR`. Should `BuiltInInteropVariantExtensions.ToObject` be changed?

Reply:
It doesn't seem to be missing a case for `VT_ERROR`. Am I missing something?

https://github.com/dotnet/runtime/blob/2567a9467daf9c429fedfc7df845f5f04316dcda/src/libraries/Common/src/System/Runtime/InteropServices/BuiltInVariantExtensions.cs#L134-L160

Reply:
It's not covered by OAVariant, but for the early bound Variant case:
https://github.com/dotnet/runtime/blob/83b0d939bedadf7d782b0b26307c2d8c1d5b76f4/src/coreclr/vm/olevariant.cpp#L4161-L4183

Certain ERROR is converted to CV_MISSING.

Reply:
> It's not covered by OAVariant, but for the early bound Variant case:

I don't understand this response. The following statement was made above at https://github.com/dotnet/runtime/pull/100176#discussion_r1543319438.

> > Specifically speaking, BuiltInInteropVariantExtensions.ToObject is missing a case of VT_ERROR. Should BuiltInInteropVariantExtensions.ToObject be changed?

Was this statement a mistake? I don't see how this is relates to reply regarding `MarshalErrorVariantOleToCom()`.